src/System.Private.CoreLib/shared/System/Text/DecoderFallback.cs

```diff
@@ -82,6 +82,116 @@ internal unsafe void InternalInitialize(byte* byteStart, char* charEnd)
+        /// </remarks>
+        internal int InternalGetFallbackCharCount(byte[] bytesUnknown)
+        {
+            // The first parameter to this method is a byte[] instead of a ROS<byte> due to an artifact of the Fallback method
+            // signature. The caller is responsible for making sure that this array is a standalone object that contains no
+            // data other than what the decoder was originally working with. It's tempting for us to take ROS<byte> and rent
+            // a buffer from the array pool, but then we have no way of letting the Fallback method know which portion of
+            // the rented array contains useful data and which portion contains garbage. We'd have to clear the rented array
+            // every time, so we may as well just make the caller take on this responsibility.
+
+            int charCount = 0;
+
+            if (Fallback(bytesUnknown, 0))
+            {
+                char thisChar;
+                while ((thisChar = GetNextChar()) != 0)
+                {
+                    if (char.IsSurrogate(thisChar))
+                    {
+                        if (!char.IsSurrogatePair(thisChar, GetNextChar()))
+                        {
+                            // Standalone high or low surrogate.
+                            throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex);
+                        }
+
+                        ++charCount; // one extra bump for the surrogate character
+                    }
+
+                    if (++charCount < 0)
+                    {
+                        Encoding.ThrowConversionOverflow(); // integer overflow
+                    }
+                }
+            }
+
+            return charCount;
+        }
+
+        /// <summary>
+        /// Determines the fallback UTF-16 data for the specified unknown bytes, then copies
+        /// the fallback data to the destination buffer.
+        /// </summary>
+        /// <returns>
+        /// The total number of chars written, or a negative value to indicate that the destination
+        /// buffer was too small to contain the fallback data.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// Thrown if the fallback generates invalid UTF-16 data (standalone surrogates).
+        /// </exception>
+        internal int InternalGetFallbackChars(byte[] bytesUnknown, Span<char> chars)
+        {
+            // See comment at top of InternalGetFallbackCharCount.
+
+            int originalCharsLength = chars.Length;
+
+            if (Fallback(bytesUnknown, 0))
+            {
+                char thisChar;
+                while ((thisChar = GetNextChar()) != 0)
+                {
+                    if (!char.IsSurrogate(thisChar))
+                    {
+                        // Standard non-surrogate BMP char.
+
+                        if (chars.IsEmpty)
+                        {
+                            goto DestinationTooSmall;
```

Comment:
What is the benefit of goto to `return -1` ?

Reply:
This originally had other logic in it that got replaced with `return -1`. I can clean it up.